### PR TITLE
fix: parse username from git url when using SSH key auth

### PR DESCRIPTION
**Pull request by chtcvl**
Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

#5147

### CODE REVIEW

File: workflow/artifacts/git/git.go [hunk 1]
```diff
@@ -27,7 +29,20 @@ type GitArtifactDriver struct {
 	InsecureIgnoreHostKey bool
 }
 
-func (g *GitArtifactDriver) auth() (func(), transport.AuthMethod, []string, error) {
+var (
+	sshURLRegex = regexp.MustCompile("^(ssh://)?([^/:]*?)@[^@]+$")
+)
+
+// IsSSHURL returns true if supplied URL is SSH URL - referenced https://github.com/argoproj/argo-cd/blob/d516f4754f4ddf99ee383452d4ff529f1cc0530c/util/git/git.go#L73
```
Comment (simster7) [inline workflow/artifacts/git/git.go:36]: No need to say where we referenced

File: workflow/artifacts/git/git.go [hunk 2]
```diff
@@ -41,7 +56,11 @@ func (g *GitArtifactDriver) auth() (func(), transport.AuthMethod, []string, erro
 		if err != nil {
 			return nil, nil, nil, err
 		}
-		auth := &ssh2.PublicKeys{User: "git", Signer: signer}
+		var sshUser string
```
Comment (simster7) [inline workflow/artifacts/git/git.go:59]: Shouldn't the default `sshUser` be `git`?
```suggestion
		sshUser := "git"
```

File: workflow/artifacts/git/git.go [hunk 3]
```diff
@@ -27,7 +29,20 @@ type GitArtifactDriver struct {
 	InsecureIgnoreHostKey bool
 }
 
-func (g *GitArtifactDriver) auth() (func(), transport.AuthMethod, []string, error) {
+var (
+	sshURLRegex = regexp.MustCompile("^(ssh://)?([^/:]*?)@[^@]+$")
+)
+
+// IsSSHURL returns true if supplied URL is SSH URL - referenced https://github.com/argoproj/argo-cd/blob/d516f4754f4ddf99ee383452d4ff529f1cc0530c/util/git/git.go#L73
+func IsSSHURL(url string) (bool, string) {
```
Comment (simster7) [inline workflow/artifacts/git/git.go:37]: Two things:
* Don't export this function (begin function name with lower case)
* Return `(string, bool)` fits convention

```suggestion
func isSSHURL(url string) (string, bool) {
```

File: workflow/artifacts/git/git.go [hunk 4]
```diff
@@ -3,9 +3,11 @@ package git
 import (
 	"errors"
 	"io/ioutil"
+	"net/url"
```
Comment (simster7) [inline workflow/artifacts/git/git.go:6]: Seems this import is not used?

**Review by simster7 — CHANGES_REQUESTED** [review]

File: go.sum [hunk 5]
```diff
@@ -1387,6 +1387,9 @@ google.golang.org/grpc v1.29.1/go.mod h1:itym6AZVZYACWQqET3MqgPpjcuV5QH3BxFS3Iji
 google.golang.org/grpc v1.31.0/go.mod h1:N36X2cJ7JwdamYAgDz+s+rVMFjt3numwzf/HckM8pak=
 google.golang.org/grpc v1.33.1 h1:DGeFlSan2f+WEtCERJ4J9GJWk15TxUi8QGagfI87Xyc=
 google.golang.org/grpc v1.33.1/go.mod h1:fr5YgcSWrqhRRxogOsw7RzIpsmvOZ6IcH4kBYTpR3n0=
+google.golang.org/grpc v1.35.0 h1:TwIQcH3es+MojMVojxxfQ3l3OF2KzlRxML2xZq0kRo8=
+google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.1.0 h1:M1YKkFIboKNieVO5DLUEVzQfGwJD30Nv2jfUgzb5UcE=
+google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.1.0/go.mod h1:6Kw0yEErY5E/yWrBtf03jp27GLLJujG4z/JK95pnjjw=
```
Comment (simster7) [inline go.sum:1392]: The changes to `go.{mod,sum}` are likely unintentional. Please either revert these files or run `make lint`

**Review by simster7 — CHANGES_REQUESTED** [review]

**Comment by simster7** [comment]
You'll need to merge with master please

**Review by simster7 — APPROVED** [review]
LGTM. @alexec Could I also ask you to take a look at this please? Two sets of eyes are better than one

File: workflow/artifacts/git/git.go [hunk 6]
```diff
@@ -27,7 +28,17 @@ type GitArtifactDriver struct {
 	InsecureIgnoreHostKey bool
 }
 
-func (g *GitArtifactDriver) auth() (func(), transport.AuthMethod, []string, error) {
+var sshURLRegex = regexp.MustCompile("^(ssh://)?([^/:]*?)@[^@]+$")
+
+func isSSHURL(url string) (string, bool) {
```
Comment (alexec) [inline workflow/artifacts/git/git.go:33]: needs tests

File: workflow/artifacts/git/git.go [hunk 7]
```diff
@@ -96,7 +111,7 @@ func (g *GitArtifactDriver) Save(string, *wfv1.Artifact) error {
 }
 
 func (g *GitArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) error {
-	closer, auth, env, err := g.auth()
+	closer, auth, env, err := g.auth(inputArtifact.Git.Repo)
```
Comment (alexec) [inline workflow/artifacts/git/git.go:114]: could pass the username here instead of whole URL

File: workflow/artifacts/git/git.go [hunk 8]
```diff
@@ -41,7 +52,11 @@ func (g *GitArtifactDriver) auth() (func(), transport.AuthMethod, []string, erro
 		if err != nil {
 			return nil, nil, nil, err
 		}
-		auth := &ssh2.PublicKeys{User: "git", Signer: signer}
+		sshUser := "git"
```
Comment (alexec) [inline workflow/artifacts/git/git.go:55]: what about adding a `GetUser(repoUrl` func?

**Review by alexec — CHANGES_REQUESTED** [review]
See comments.

**Comment by simster7** [comment]
@chtcvl Bump! Any interest in continuing this PR?

**Comment by chtcvl** [comment]
> @chtcvl Bump! Any interest in continuing this PR?

yes, will address the requested changes shortly

File: workflow/artifacts/git/git.go (hunk 6 above)
Comment (chtcvl) [reply workflow/artifacts/git/git.go]: tests added

File: workflow/artifacts/git/git.go (hunk 8 above)
Comment (chtcvl) [reply workflow/artifacts/git/git.go]: added `GetUser` function

File: workflow/artifacts/git/git.go (hunk 7 above)
Comment (chtcvl) [reply workflow/artifacts/git/git.go]: updated to pass username in stead of repo url

File: workflow/artifacts/git/git.go [hunk 9]
```diff
@@ -31,7 +32,18 @@ type ArtifactDriver struct {
 
 var _ common.ArtifactDriver = &ArtifactDriver{}
 
-func (g *ArtifactDriver) auth() (func(), transport.AuthMethod, []string, error) {
+var sshURLRegex = regexp.MustCompile("^(ssh://)?([^/:]*?)@[^@]+$")
+
+func GetUser(url string) (string, bool) {
```
Comment (alexec) [inline workflow/artifacts/git/git.go:37]: Bool is always ignored, so you can remove it

**Review by alexec — CHANGES_REQUESTED** [review]

File: workflow/artifacts/git/git.go (hunk 9 above)
Comment (chtcvl) [reply workflow/artifacts/git/git.go]: remove ignored bool output

**Review by alexec — APPROVED** [review]

**Pull request merged**